### ISSUE TITLE
What to do with Debug

### ISSUE THREAD

**Issue by shadow-fa** [Raptoer/F3#162]
Branching off of this discussion https://github.com/ferstaberinde/F3/issues/514 ,
I'd like to discuss the future of debug in FA3.

### Current state
Currently F(A)3 uses a mixture of `diag_log`,`systemChat`,`globalChat`,`sideChat` and `BIS_fnc_error` for debug/error messages with an inconsistent structure where the messages sometimes don't even include the component/script name.

Some components create a lot of spam in (side)Chat.

### Idea
Since we don't to include some `macro.hpp` in _every_ other script file,
I think the best idea would be to have a single debug function.

This could look something like:
```sqf
[format ["message %1", _something], __FILE__, __LINE__] call f_fnc_debug;
```
where `file` and `line` would be optional parameters.
(Note: In case of precompiled functions we have the (parent-)script name anyway, but not for other scripts, spawned code, EH-code, etc.)
This function would then print the message to systemChat(?) and diag_log.
There could also be a `f_fnc_debugError` function which also calls `BIS_fnc_error` and will always print the message even if debug is disabled.

### Advantages
The advantage of this would be, that we can make sure that everything is also logged in the `.rpt` file, and we can add any necessary checks inside this function (e.g. only print messages from component X).
Additionally missionmakers could use these functions for their custom scripts.

### Issues
This does not solve the problem that the chat is spammed with debug messages, however it makes it easier to look in the `.rpt` file to read and copy every debug message. We could also add a parameter option to select from `sideChat` or `diag_log` to alleviate this problem.

There is also an issue of locality: Do we want to log debug messages from code that runs on other machines as well? (This could lead to network issues if the function is called too often)

### Discussion

I have some prototype code to test what's possible without macros, but ultimately these are just ideas.
I do think that the current state of f3 debug is a mess and _something_ should be done.
I'd like to invite everyone to discuss the debug situtation to find a solution.

**Comment by SamLex** [comment]
This is definitely a 3.6.0 thing since requires a lot of work, but here is my current thinking on debug and error logging. There should be two types of message, errors and debug, and they should all be logged in the same way. Errors should be generated when *something* has gone wrong, or was misconfigured, or doesn't make sense. This should be irrespective of if the function then continues with a default, just rolls with it, or crashes and burns. Debug messages are for debugging (duh!) by logging important information (e.g. function parameters) and decisions made. Thus I propose the following:

### f/common/logging
A module/package to encapsulate logging behaviour with the following functions:

- f_fnc_setDebugMode [enabled: boolean] - simply sets f_var_debugMode = enabled. Would be called from init.sqf with value of f_param_debugMode. Keeps parameters out of functions and allows additional behaviour than just a flag in future
- f_fnc_logError [module: string, message: string, fparam: [any], file: string (optional), line: int (optional)] - would *log* (see below) an error for the specified module containing the formatted message, and file and line number if provided
  - module - the name of the module the error happened in
  - message - the error message, possibly containing formatting placeholders
  - fparam - array to fill placeholders from
  - file - optional file name from __FILE__
  - line - optional line number from __LINE__
- f_fnc_logDebug (arguments as f_fnc_logError) - If f_var_debugMode == false, does nothing. Otherwise would *log* (see below) a debug message for the specified module containing the formatted message, and file and line number if provided

#### Logging
I think all messages should be logged to the RPT file (and console for the case of the server and HC) with `diag_log`. In addition, `createDiarySection` and `createDiaryRecord` should be used on clients with an interface. An example of what this might look like is below.
![arma 3 screenshot 2018 11 21 - 16 37 07 91](https://user-images.githubusercontent.com/733683/48856774-3fc90b00-edaf-11e8-9203-f8c55d8d3706.png)
Created with:
```
player createDiarySubject ["fa3_log", "FA3 Log"];
player createDiaryRecord ["fa3_log", ["assignGear","[ERROR] some assignGear error that might be too long to fit in the title"]];
player createDiaryRecord ["fa3_log", ["assignGear","[ERROR] some other assignGear error that might also be too long to fit in the title"]];
player createDiaryRecord ["fa3_log", ["assignGear","[DEBUG] lets have a debug message for a change"]];
player createDiaryRecord ["fa3_log", ["assignGear","[ERROR] back to errors"]];
player createDiaryRecord ["fa3_log", ["authorisedCrew","[ERROR] back to errors"]];
```


A top-level diary section "FA3 Log" would be created for all log messages. A record within this would be lazily created for each module as message arrive. Messages would be added to each record by repeated calls to `createDiaryRecord`. In addition, interface-less clients should send their messages to those with interfaces using `remoteExec` where additional sections would be lazily added ("FA3 Log - Server" for server and "FA3 Log - HC" for HC).

The advantage of this approach is that messages are persistent and easy to read throughout a mission (diary) and afterwards (RPT), unlike chat which tails off and can be hard to follow. A potential disadvantage is that it won't be obvious when an error occurs but this could also be seen as an advantage, depending on exactly what went wrong.
